File: 20-actors/niyaku/methods/test_stow_plan.py
"""Tests for stow_plan — stowage slotting + discharge sequencing."""

import pytest

from stow_plan import (
    Container,
    Slot,
    StowError,
    build_stow_plan,
    discharge_sequence,
    validate_no_rehandle,
)


def _rot_index(rotation):
    return {p: i for i, p in enumerate(rotation)}


def test_simple_plan_places_all():
    rotation = ["SHA", "SIN", "ROT"]
    # weight must agree with discharge order for a single stack to be feasible:
    # earliest-discharge (SHA) sits on top, so it must be the lightest.
    boxes = [
        Container("A", 22.0, "ROT"),  # last off → heaviest → bottom
        Container("B", 18.0, "SIN"),
        Container("C", 14.0, "SHA"),  # first off → lightest → top
    ]
    plan = build_stow_plan(boxes, rotation, bays=1, rows=1, tiers=3)
    assert set(plan.assignments) == {"A", "B", "C"}
    # later-discharge ROT (A) must be at the bottom (tier 0)
    assert plan.slot_of("A").tier < plan.slot_of("B").tier
    assert plan.slot_of("B").tier < plan.slot_of("C").tier
    box_port = {b.box_id: b.discharge_port for b in boxes}
    assert validate_no_rehandle(plan, _rot_index(rotation), box_port)


def test_weight_on_top_not_violated():
    rotation = ["P1"]
    boxes = [
        Container("light", 5.0, "P1"),
        Container("heavy", 25.0, "P1"),
    ]
    plan = build_stow_plan(boxes, rotation, bays=1, rows=1, tiers=2)
    # heavy must end up below light (lower tier)
    assert plan.slot_of("heavy").tier < plan.slot_of("light").tier


def test_capacity_exceeded_raises():
    rotation = ["P1"]
    boxes = [Container(f"b{i}", 10.0, "P1") for i in range(5)]
    with pytest.raises(StowError):
        build_stow_plan(boxes, rotation, bays=1, rows=1, tiers=4)


def test_reefer_only_in_reefer_rows():
    rotation = ["P1"]
    boxes = [Container("r", 10.0, "P1", reefer=True)]
    plan = build_stow_plan(boxes, rotation, bays=1, rows=2, tiers=1, reefer_rows=[1])
    assert plan.slot_of("r").row == 1


def test_reefer_infeasible_when_no_reefer_row():
    rotation = ["P1"]
    boxes = [Container("r", 10.0, "P1", reefer=True)]
    with pytest.raises(StowError):
        build_stow_plan(boxes, rotation, bays=1, rows=1, tiers=1, reefer_rows=[])


def test_hazmat_segregation_separates_classes():
    rotation = ["P1"]
    boxes = [
        Container("flam", 10.0, "P1", hazmat="3"),
        Container("oxid", 10.0, "P1", hazmat="5.1"),
    ]
    # two classes cannot share a column; need ≥2 columns
    plan = build_stow_plan(boxes, rotation, bays=2, rows=1, tiers=2)
    assert (plan.slot_of("flam").bay, plan.slot_of("flam").row) != (
        plan.slot_of("oxid").bay, plan.slot_of("oxid").row
    )
    # only one column ⇒ infeasible
    with pytest.raises(StowError):
        build_stow_plan(boxes, rotation, bays=1, rows=1, tiers=2)


def test_unknown_port_raises():
    with pytest.raises(StowError):
        build_stow_plan([Container("x", 1.0, "ZZZ")], ["P1"], 1, 1, 1)


def test_empty_rotation_raises():
    with pytest.raises(StowError):
        build_stow_plan([], [], 1, 1, 1)


def test_discharge_sequence_top_first():
    rotation = ["P1"]
    boxes = [Container(f"b{i}", 10.0 - i, "P1") for i in range(3)]
    plan = build_stow_plan(boxes, rotation, bays=1, rows=1, tiers=3)
    seq = discharge_sequence(plan, "P1")
    # top tier discharged first
    tiers = [plan.slot_of(bid).tier for bid in seq]
    assert tiers == sorted(tiers, reverse=True)


def test_slot_key():
    assert Slot(1, 2, 3).key() == (1, 2, 3)

File: 20-actors/niyaku/methods/test_stow_plan.clj
#!/usr/bin/env bb
;; Working Clojure port of methods/test_stow_plan.py.
(ns niyaku.methods.test-stow-plan
  "Tests for stow_plan — stowage slotting + discharge sequencing.

  Run:  bb --classpath 20-actors 20-actors/niyaku/methods/test_stow_plan.clj"
  (:require [niyaku.methods.stow-plan :as st]
            [clojure.test :refer [deftest is run-tests]]))

(defn- rot-index [rotation] (into {} (map-indexed (fn [i p] [p i]) rotation)))

(deftest simple-plan-places-all
  (let [rotation ["SHA" "SIN" "ROT"]
        boxes [(st/container "A" 22.0 "ROT")    ; last off → heaviest → bottom
               (st/container "B" 18.0 "SIN")
               (st/container "C" 14.0 "SHA")]   ; first off → lightest → top
        plan (st/build-stow-plan boxes rotation 1 1 3)]
    (is (= (set (keys (:assignments plan))) #{"A" "B" "C"}))
    (is (< (:tier (st/slot-of plan "A")) (:tier (st/slot-of plan "B"))))
    (is (< (:tier (st/slot-of plan "B")) (:tier (st/slot-of plan "C"))))
    (let [box-port (into {} (map (juxt :box-id :discharge-port) boxes))]
      (is (st/validate-no-rehandle plan (rot-index rotation) box-port)))))

(deftest weight-on-top-not-violated
  (let [plan (st/build-stow-plan [(st/container "light" 5.0 "P1")
                                  (st/container "heavy" 25.0 "P1")] ["P1"] 1 1 2)]
    (is (< (:tier (st/slot-of plan "heavy")) (:tier (st/slot-of plan "light"))))))

(deftest capacity-exceeded-raises
  (is (thrown? clojure.lang.ExceptionInfo
               (st/build-stow-plan (map #(st/container (str "b" %) 10.0 "P1") (range 5))
                                   ["P1"] 1 1 4))))

(deftest reefer-only-in-reefer-rows
  (let [plan (st/build-stow-plan [(st/container "r" 10.0 "P1" :reefer true)]
                                 ["P1"] 1 2 1 :reefer-rows [1])]
    (is (= (:row (st/slot-of plan "r")) 1))))

(deftest reefer-infeasible-when-no-reefer-row
  (is (thrown? clojure.lang.ExceptionInfo
               (st/build-stow-plan [(st/container "r" 10.0 "P1" :reefer true)]
                                   ["P1"] 1 1 1 :reefer-rows []))))

(deftest hazmat-segregation-separates-classes
  (let [boxes [(st/container "flam" 10.0 "P1" :hazmat "3")
               (st/container "oxid" 10.0 "P1" :hazmat "5.1")]
        plan (st/build-stow-plan boxes ["P1"] 2 1 2)]
    (is (not= [(:bay (st/slot-of plan "flam")) (:row (st/slot-of plan "flam"))]
              [(:bay (st/slot-of plan "oxid")) (:row (st/slot-of plan "oxid"))]))
    (is (thrown? clojure.lang.ExceptionInfo (st/build-stow-plan boxes ["P1"] 1 1 2)))))

(deftest unknown-port-raises
  (is (thrown? clojure.lang.ExceptionInfo
               (st/build-stow-plan [(st/container "x" 1.0 "ZZZ")] ["P1"] 1 1 1))))

(deftest empty-rotation-raises
  (is (thrown? clojure.lang.ExceptionInfo (st/build-stow-plan [] [] 1 1 1))))

(deftest discharge-sequence-top-first
  (let [boxes (map #(st/container (str "b" %) (- 10.0 %) "P1") (range 3))
        plan (st/build-stow-plan boxes ["P1"] 1 1 3)
        seq* (st/discharge-sequence plan "P1")
        tiers (map #(:tier (st/slot-of plan %)) seq*)]
    (is (= tiers (sort > tiers)))))   ; top tier discharged first

(deftest weight-on-top-forces-new-column
  ;; a heavy early-discharge box cannot sit on a light late-discharge box → fresh column
  (let [plan (st/build-stow-plan [(st/container "late_light" 10.0 "P1")
                                  (st/container "early_heavy" 20.0 "P0")] ["P0" "P1"] 2 1 2)]
    (is (not= (:bay (st/slot-of plan "late_light")) (:bay (st/slot-of plan "early_heavy"))))))

(deftest validate-no-rehandle-detects-violation
  ;; a hand-built plan with a later-discharge box stacked above an earlier one → false
  (let [plan {:assignments {"below_early" (st/slot 0 0 0)    ; P0 (early) at bottom — bad
                            "above_late" (st/slot 0 0 1)}    ; P1 (late) on top — buries early
              :rotation []}]
    (is (false? (st/validate-no-rehandle plan {"P0" 0 "P1" 1}
                                         {"below_early" "P0" "above_late" "P1"})))))

(deftest slot-key-test
  (is (= (st/slot-key (st/slot 1 2 3)) [1 2 3])))

(deftest stow-errors-are-typed
  (is (try (st/build-stow-plan [] [] 1 1 1) false
           (catch clojure.lang.ExceptionInfo e (st/stow-error? e)))))

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (run-tests 'niyaku.methods.test-stow-plan)]
    (System/exit (if (zero? (+ fail error)) 0 1))))
